qhana_plugin_registry/db/models/templates.py

```diff
@@ -75,6 +76,41 @@ class UiTemplate(IdMixin, NameDescriptionMixin, ExistsMixin):
         },
     )
 
+    @classmethod
+    def get_by_name(cls, name: str) -> "UiTemplate":
+        q = select(cls).filter(cls.name == name)
+        try:
+            found_template: UiTemplate = DB.session.execute(q).scalar_one_or_none()
+        except MultipleResultsFound:
+            found_template = DB.session.execute(q).first()[0]
+        return found_template
+
+    @classmethod
+    def from_json(cls, template_dict: dict) -> "UiTemplate":
+        template = UiTemplate(
+            name=template_dict["name"],
+            description=template_dict["description"],
+            tags=TemplateTag.get_or_create_all(template_dict["tags"]),
+            tabs=[],  # FIXME tabs are added later since they need the template
+        )
+        return template
+
+    @classmethod
+    def get_or_create_from_json(cls, template_dict: dict) -> "UiTemplate":
+        from qhana_plugin_registry.tasks.plugin_filter import apply_filter_for_tab
+
+        # only checks if template with same name exists
+        template: Optional[UiTemplate] = cls.get_by_name(template_dict["name"])
+        if template is None:
+            template = cls.from_json(template_dict)
+            for tab in template_dict["tabs"]:
+                tab["template"] = template
+            template.tabs = TemplateTab.get_or_create_all(template_dict["tabs"])
+            for tab in template.tabs:
+                apply_filter_for_tab(tab.id)
+            DB.session.add(template)
```

Comment:
The tabs aren't added in the `apply_filter_for_tab` function. Now, they're added just before it:

https://github.com/UST-QuAntiL/qhana-plugin-registry/blob/7ad8c96e3fe29aff8fb49790e4cdbd2f269b84e7/qhana_plugin_registry/db/models/templates.py#L278-L280

Reply:
Ideally you only need one commit for the template and tab creation. It should look something like this (with some fictional function calls):

```python
template = create_template(template_data)
template.tabs = [create_tab(tab_data, template) for tab_data in template_data["tabs"]]
DB.session.add(template)  # may be omitted if create_template adds template to DB
for tab in template.tabs:
    # avoid loading an object from DB if it is already in memory
    # this also removes the need for commiting the data to the DB
    tab.plugins = list(evaluate_plugin_filter(tab.plugin_filter))
    DB.session.add(tab)  # may be omitted if create_tab adds tab to DB
DB.session.commit()  # only having one commit for the entire session ensures that the template gets loaded consistently or not at all!
```

If you are in a situation where you need stuff to be in the database but do not want to commit the session you can use `DB.session.flush()`. But in this case this can be entirely avoided.